notification-server/src/google_cloud.rs

```diff
@@ -1,57 +1,162 @@
-use crate::models::NotificationInformation;
-use anyhow::anyhow;
+use crate::{
+    google_cloud::NotificationError::{
+        AuthenticationError, ClientError, InvalidArgumentError, ServerError, UnregisteredError,
+    },
+    models::NotificationInformation,
+};
 use backoff::{future::retry, ExponentialBackoff};
-use gcp_auth::{CustomServiceAccount, TokenProvider};
-use reqwest::Client;
+use gcp_auth::TokenProvider;
+use reqwest::{Client, StatusCode};
 use serde_json::json;
-use std::{collections::HashMap, path::PathBuf};
+use std::{
+    collections::HashMap,
+    fmt::{Display, Formatter},
+};
 
 const SCOPES: &[&str; 1] = &["https://www.googleapis.com/auth/firebase.messaging"];
 
-pub struct GoogleCloud {
+#[derive(Debug, PartialEq)]
+pub enum NotificationError {
+    ServerError(String),
+    ClientError(String),
+    InvalidArgumentError,
+    AuthenticationError(String),
+    UnregisteredError,
+}
+
+impl Display for NotificationError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ClientError(msg) => write!(f, "Client Error: {}", msg),
+            ServerError(msg) => write!(f, "Server Error: {}", msg),
+            InvalidArgumentError => write!(f, "Device token had an invalid format"),
+            UnregisteredError => write!(f, "Device token has been unregistered"),
+            AuthenticationError(msg) => write!(f, "Authentication error: {}", msg),
+        }
+    }
+}
+
+impl std::error::Error for NotificationError {}
```

Comment:
Using the `thiserror` crate to define custom errors would be good:
```suggestion
#[derive(Debug, PartialEq, thiserror::Error)]
pub enum NotificationError {
    #[error("Client Error: {0}")] // and so on...
    ServerError(String),
    ClientError(String),
    InvalidArgumentError,
    AuthenticationError(String),
    UnregisteredError,
}
```

Reply:
That is indeed a much neater solution